02-express-tutorial/app.js

```diff
@@ -1 +1,63 @@
-console.log('Express Tutorial')
+const express = require("express");
+const { products } = require("./data");
+
+//invoke express
+const app = express();
+
+// setup static and middleware
+app.use(express.static("./public"));
+
+app.get("/api/v1/test", (req, res) => {
+  res.json({ message: "It worked!" });
+});
+
+app.get("/api/v1/products", (req, res) => {
+  res.json(products);
+});
+
+app.get("/api/v1/products/:productID", (req, res) => {
+  const idToFind = parseInt(req.params.productID);
+  const product = products.find((p) => p.id === idToFind);
+
+  if (!product) {
+    return res.status(404).send({ message: "That product was not found." });
+  }
+
+  return res.json(product);
+});
+//search 
+app.get('/api/v1/query', (req, res) => {
+  const { search, limit, regex, price } = req.query;
+  let sortedProducts = [...products];
+
+  if (regex) {
+    const searchByRegex = new RegExp(regex, "i");
+    sortedProducts = sortedProducts.filter((product) =>
+      searchByRegex.test(product.name)
+    );
+  } else if (search) {
+    sortedProducts = sortedProducts.filter((product) => {
+      return product.name.toLowerCase().startsWith(search.toLowerCase());
+    });
+  }
+
+  if (limit) {
+    sortedProducts = sortedProducts.slice(0, Number(limit));
+  }
+  if (price) {
+    sortedProducts = sortedProducts.filter((product) => product.price >= price);
+  }
+  if (sortedProducts.length < 1) {
```

Comment:
# Required Change
[Inconsistent Data Schema]
This code returns 2 different json schemas (aka: layouts, data models, etc.) depending on whether a matching product was found or not. This anyone calling this API has to take into account that the data returned could come in two formats. 
The two schemas are:
```js
{
  "success": bool,
  "data": Array[Product]
}
```
```js
{
  Array[Product]
}
```

The first occurs when there are no products returned, and the second when products are found. The first follows better API conventions and should be the one you stick with. The reason this is better in general is because it allows for "extensibility". For instance, in the future you may want to add a `total_count` or `more` field to the response to tell a caller that their results are cutoff due to the `limit` parameter. (Calling code may need to know this if they want to give the user the opportunity to click a more button or something.) With the first schema its easy to add that, and existing code using the API won't need to change. With the second schema, there's no straightforward way to add it that doesn't break the code for existing callers.

Reply:
Yes, It makes sense.  Thank you for detailed information.

Reply:
The requested changes were implemented. Please review them.